cranelift/interpreter/src/step.rs

```diff
@@ -484,6 +485,59 @@ where
         }
         Opcode::SymbolValue => unimplemented!("SymbolValue"),
         Opcode::TlsValue => unimplemented!("TlsValue"),
+        Opcode::HeapLoad => {
+            if let InstructionData::HeapLoad { heap_imm, arg, .. } = inst {
+                let HeapImmData {
+                    flags,
+                    heap,
+                    offset,
+                } = state.get_current_function().dfg.heap_imms[heap_imm];
+                let offset = i128::from(u32::from(offset));
+                let addr_ty = state.get_current_function().dfg.value_type(arg);
+                let index = state.get_value(arg).unwrap().into_int().unwrap();
+                let load_ty = inst_context.controlling_type().unwrap();
+                let load_size = i128::from(load_ty.bytes());
+                assign_or_memtrap(AddressSize::try_from(addr_ty).and_then(|addr_size| {
+                    let heap_address = index.saturating_add(offset).saturating_add(load_size);
```

Comment:
Why `saturating_add` here rather than e.g. `checked_add` with handling for the `None` case? I guess because these are `i128`s and not expected to actually saturate? If so could we do `.checked_add(...).expect("128-bit math should not overflow here")` or somesuch?

Reply:
Mostly because error handling in the interpreter is pretty weird and there is no easy way to turn arbitrary `impl Error` types into traps. But saturating should be correct and seems less wordy and involved than the checked adds, to me.

Reply:
I think I'd actually prefer the checked_add/expect sequence: saturating_add is explicitly *incorrect* in general for some values allowed by the types (close to the limit), and is a suspicious and unexpected thing to see in an address computation. It's correct here because of the ranges involved, but those invariants aren't implicit and not obvious to the reader at all. If we're comfortable saying that the saturation should never actually happen (in the current version), then we should be equally happy to say that the expect should never actually fail (in the proposed version), except in the latter case we'll know if we're wrong.